lib/aws/cognito_srp.rb

```diff
@@ -117,6 +119,10 @@ def authenticate
 
       auth_response = @aws_client.respond_to_auth_challenge(params)
 
+      if auth_response.challenge_name == SOFTWARE_TOKEN_MFA || auth_response.challenge_name == SMS_MFA
+        return auth_response
+      end
```

Comment:
How about using an exception here instead of returning `auth_response`? I think it would have a couple benefits:

1. If you forget to check for `resp.challenge_name == 'SOFTWARE_TOKEN_MFA'` manually then you get a meaningful error, which is important as otherwise your code would get stuck not knowing how to proceed anyway
2. An exception would allow for cleaner handling code

E.g., these lines would change to:

```suggestion
        if auth_response.challenge_name == SOFTWARE_TOKEN_MFA
          raise SoftwareTokenMfaRequired, auth_response
        elsif auth_response.challenge_name == SMS_MFA
          raise SmsMfaRequired, auth_response
        end
```

With this in `lib/aws/cognito_srp/errors.rb`:

```ruby
class MfaRequired < Error
  attr_reader :auth_response

  def initialize(auth_response)
    @auth_response = auth_response
  end
end
class SoftwareTokenMfaRequired < MfaRequired; end
class SmsMfaRequired < MfaRequired; end
```

Then you could catch it like:

```ruby
begin
  resp = aws_srp.authenticate
rescue Aws::CognitoSrp::SoftwareTokenMfaRequired => e
  user_code = get.chomp
  resp = aws_srp.respond_to_auth_challenge_mfa(
    e.auth_challenge.challenge_name,
    e.auth_challenge.session,
    user_code
  )
end
```

Or perhaps the exception class could provide a convenience method to respond to the MFA challenge directly without having to manually pass the parameters, resulting in cleaner handling code:

```ruby
begin
  resp = aws_srp.authenticate
rescue Aws::CognitoSrp::SoftwareTokenMfaRequired => e
  user_code = get.chomp
  resp = e.respond(user_code)
end
```

Please let me know what you think of this idea. I could implement it if you prefer, but I'd like your input first.

Reply:
@pilaf 
Thank you for reviewing and sorry for my response is late.

I feel uncomfortable with raising exception.
Because:

1. aws sdk does not raise exception when user and password are collect.
2. I think the normal code should not be in rescue block.
3. I feel uncomfortable with exception although the first authentication (user and password authentication) step is success

Reply:
@pilaf 
What do you think about my above comment?